File: Solutions/Python/unique_paths_ii.py
"""
    [MEDIUM]
    63. Unique Paths II

    Concepts:
    - matrix
    - dynamic programming

    Stats:
        Runtime | 3 ms      [Beats 26.13%]
        Memory  | 18.23 MB  [Beats 7.44%]
"""

class Solution:
    def uniquePathsWithObstacles(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dp = defaultdict(int)

        if (m, n) == (1, 1):
            return 0 if grid[0][0] else 1

        if grid[0][0]:
            return 0

        def dfs(i, j):
            nonlocal dp

            if (i, j) == (m - 1, n - 1):
                dp[(i, j)] = 1
                return dp[(i, j)]

            if (i, j) in dp:
                return dp[(i, j)]

            if i + 1 < m and not grid[i+1][j]:
                dp[(i, j)] += dfs(i+1, j)
            if j + 1 < n and not grid[i][j+1]:
                dp[(i, j)] += dfs(i, j+1)

            return dp[(i, j)]

        return dfs(0, 0)

File: Solutions/Cpp/unique_paths_ii.cpp
/*
    [MEDIUM]
    63. Unique Paths II

    Concepts:
    - matrix
    - dynamic programming

    Stats:
        Runtime | 0 ms      [Beats 100%]
        Memory  | 15.80 MB  [Beats 6.57%]
*/

class Solution {
public:
    int m, n;

    inline int coords_to_idx(const int i, const int j) {
        return (i * n) + (j % n);
    }

    int uniquePathsWithObstacles(vector<vector<int>>& grid) {
        m = grid.size(), n = grid[0].size();
        std::unordered_map<int, int> dp = {};

        if (m == 1 && n == 1)
            return !grid[0][0];

        if (grid[0][0])
            return 0;

        std::function<int(int, int)> dfs;
        dfs = [&](int i, int j){
            int idx = coords_to_idx(i, j);

            if (i == m - 1 && j == n - 1) {
                dp[idx] = 1;
                return dp[idx];
            }

            if (dp.contains(idx))
                return dp[idx];

            if (i + 1 < m && !grid[i+1][j])
                dp[idx] += dfs(i + 1, j);
            if (j + 1 < n && !grid[i][j+1])
                dp[idx] += dfs(i, j + 1);

            return dp[idx];
        };

        return dfs(0, 0);
    }
};